src/main/kotlin/org/jetbrains/research/testspark/tools/llm/generation/JUnitTestsAssembler.kt

```diff
@@ -78,7 +81,14 @@ class JUnitTestsAssembler(
         return testSuite
     }
 
-    private fun createTestSuiteParser(packageName: String, jUnitVersion: JUnitVersion): TestSuiteParser {
-        return JUnitTestSuiteParser(packageName, jUnitVersion)
+    private fun createTestSuiteParser(
+        packageName: String,
+        jUnitVersion: JUnitVersion,
+        language: Language,
+    ): TestSuiteParser {
+        return when (language) {
+            Language.Java -> JavaJUnitTestSuiteParser(packageName, jUnitVersion, javaImportPattern)
+            Language.Kotlin -> KotlinJUnitTestSuiteParser(packageName, jUnitVersion, javaImportPattern)
+        }
```

Comment:
**_[Any changes coming from this comment should be addressed in a different PR]_**

I thought that, by language, we select an implementation of the `TestsAssembler` and its underlying' TestSuiteParser` in runtime (see image).

<img width="868" alt="Screenshot 2024-07-12 at 16 44 05" src="https://github.com/user-attachments/assets/d4132a12-4a78-4d40-beb3-ab14c5ab6bc4">

In the image, we have a factory that produces an instance of `TestsAssembler` depending on the language (and, in most general case, the testing framework selected):

```kotlin
class TestsAssemblerFactory {
  getAssembler(language: Langauge) {
    return when(language) {
      Java -> JavaTestsAssembler(parser = JavaJUnitTestSuiteParser(junitVersion))
      Kotlin -> KotlinTestsAssembler(parser = KotlinJUnitTestSuiteParser(junitVersion))
    }
  }
}
```

Notice that `TestsAssembler` does not know about the testing framework: it's the responsibility of the `TestSuiteParser` to figure out how to parse the test suite presented in every concrete testing platform:

Example with `JavaTestsAssembler`:
```kotlin
class JavaTestsAssembler : TestsAssembler {
  constructor(parser: TestSuiteParser) {
    super(parser);
  }

  override fun assembleTestSuite(pkgName): TestSuite? {
   // parser already knows about the JUnit version, as it was created with its
    val testSuite = this.parser(pkgName, super.getContent())
  }
}
```

Such structure, in theory, allows extension with the support of an arbitrary testing framework for any programming language.

The current implementation requires propagating the instance of `Langauge` into many layers, which complicates the APIs of the classes involved + hard-codes test suite parsers.


Some technical issues might arise when migrating to this structure. Need to discuss @pderakhshanfar, @Frosendroska, @arksap2002.

Reply:
I really like this comment. I was thinking about this problem while doing the next PR, so lets merge ICTL907,908 and I will work on that, or probably address it in the 908 already, because there I do some related changes

Reply:
We already know the JUnitVersion, but we can not call it in the `assembleTestSuite`, because now it is framework agnostic, and can not call it in `parseTestSuite`, because it is the interface function for all the Language/Framework pairs. What is your suggestion then?

Reply:
I do not think that this change is relevant before we start making the TestSpark "testing framework agnostic". The thing is that currently we rely a lot on the Junit: use the version of it and some specific semantics. This is the reason why JUnitTestAssembler exist -- the algorithm for assembling test suites is the same for Kotlin and Java (except test suit parsing) so we will have a lot of duplicates. 

For example, the JavaTestsAssembler will look like this

```
class JavaJUnitTestsAssembler(
    private val parser: JUnitTestSuiteParser,
    val project: Project,
    val indicator: CustomProgressIndicator,
    val generationData: TestGenerationData,
) : TestsAssembler() {

    private val log: Logger = Logger.getInstance(this.javaClass)

    private var lastTestCount = 0

    /**
     * Receives a response text and updates the progress bar accordingly.
     *
     * @param text part of the LLM response
     */
    override fun consume(text: String) {
        if (text.isEmpty()) return

        // Collect the response and update the progress bar
        super.consume(text)
        updateProgressBar()
    }

    private fun updateProgressBar() {
        val generatedTestsCount = super.getContent().split("@Test").size - 1

        if (lastTestCount != generatedTestsCount) {
            indicator.setText(PluginMessagesBundle.get("generatingTestNumber") + generatedTestsCount)
            lastTestCount = generatedTestsCount
        }
    }

    override fun assembleTestSuite(
        packageName: String,
    ): TestSuiteGeneratedByLLM? {
        val testSuite = parser.parseTestSuite(super.getContent())
        // save RunWith
        if (testSuite?.runWith?.isNotBlank() == true) {
            generationData.runWith = testSuite.runWith
            generationData.importsCode.add(parser.getJUnitVersion().runWithAnnotationMeta.import)
        } else {
            generationData.runWith = ""
            generationData.importsCode.remove(parser.getJUnitVersion().runWithAnnotationMeta.import)
        }

        // save annotations and pre-set methods
        generationData.otherInfo = testSuite?.otherInfo ?: ""

        // logging generated test cases if any
        testSuite?.testCases?.forEach { testCase -> log.info("Generated test case: $testCase") }
        return testSuite
    }
}
```

and it will be same as KotlintestsAssembler, because we use the Unit not only for parsing but also for adding other important parts to the tests

Reply:
In this case, let's keep the proposed change for sometime later.